Sources/_StringProcessing/Engine/Processor.swift

```diff
@@ -339,6 +340,107 @@ extension Processor {
     }
   }
 
+  mutating func regexAssert(
+    by kind: AST.Atom.AssertionKind,
+    _ anchorsMatchNewlines: Bool,
+    _ usesSimpleUnicodeBoundaries: Bool,
+    _ usesASCIIWord: Bool,
+    _ semanticLevel: MatchingOptions.SemanticLevel
+  ) throws -> Bool {
+    // Future work: Optimize layout and dispatch
+    
+    // FIXME: Depends on API model we have... We may want to
+    // think through some of these with API interactions in mind
+    //
+    // This might break how we use `bounds` for both slicing
+    // and things like `firstIndex`, that is `firstIndex` may
+    // need to supply both a slice bounds and a per-search bounds.
+    switch kind {
+    case .startOfSubject: return currentPosition == subjectBounds.lowerBound
+
+    case .endOfSubjectBeforeNewline:
+      if currentPosition == subjectBounds.upperBound { return true }
+      switch semanticLevel {
+      case .graphemeCluster:
+        return input.index(after: currentPosition) == subjectBounds.upperBound
+         && input[currentPosition].isNewline
+      case .unicodeScalar:
+        return input.unicodeScalars.index(after: currentPosition) == subjectBounds.upperBound
+         && input.unicodeScalars[currentPosition].isNewline
+      }
+
+    case .endOfSubject: return currentPosition == subjectBounds.upperBound
+
+    case .resetStartOfMatch:
+      // FIXME: Figure out how to communicate this out
+      throw Unsupported(#"\K (reset/keep assertion)"#)
+
+    case .firstMatchingPositionInSubject:
+      // TODO: We can probably build a nice model with API here
+
+      // FIXME: This needs to be based on `searchBounds`,
+      // not the `subjectBounds` given as an argument here
+      // (Note: the above fixme was in reference to the old assert function API.
+      //   Now that we're in processor, we have access to searchBounds)
+      return false
+
+    case .textSegment: return input.isOnGraphemeClusterBoundary(currentPosition)
+
+    case .notTextSegment: return !input.isOnGraphemeClusterBoundary(currentPosition)
+
+    case .startOfLine:
+      // FIXME: Anchor.startOfLine must always use this first branch
+      // The behavior of `^` should depend on `anchorsMatchNewlines`, but
+      // the DSL-based `.startOfLine` anchor should always match the start
+      // of a line. Right now we don't distinguish between those anchors.
```

Comment:
I think this should ideally be resolved at compilation time, since options are compilation options and not run-time options

Reply:
I think what the fixme is saying is that the DSLTree doesn't have a way to represent them differently right now, so we have no way of differentiating between them

Reply:
Fixed in https://github.com/apple/swift-experimental-string-processing/pull/580